src/persistence/in_memory.rs

```diff
@@ -0,0 +1,66 @@
+use std::collections::BTreeMap;
+
+use parking_lot::Mutex;
+
+use crate::auction::ItemId;
+
+use super::*;
+
+pub fn new() -> Persistence {
+    Persistence(Arc::new(InMemoryPersistence::default()))
+}
+
+/// Fake in-memory persistence.
+///
+/// Useful for unit-tests.
+#[derive(Debug, Default)]
+struct InMemoryPersistence(RwLock<BTreeMap<ItemId, AuctionBiddingState>>);
+
+impl PersistenceImpl for InMemoryPersistence {
+    fn get_connection(self: Arc<Self>) -> Result<Connection> {
+        let res = Connection(Box::new(InMemoryConnection { lock: self.clone() }));
+        Ok(res)
+    }
+}
+
+#[derive(Default, Debug)]
+struct InMemoryConnection {
+    lock: Arc<InMemoryPersistence>,
+}
+
+impl ConnectionImpl for InMemoryConnection {
+    fn start_transaction<'a>(&'a mut self) -> Result<Transaction<'a>> {
+        let res = Transaction(Box::new(InMemoryTransaction {
+            lock_guard: self.lock.0.write().expect("lock to work"),
+        }));
+        Ok(res)
+    }
+}
+
+#[derive(Debug)]
+struct InMemoryTransaction<'a> {
+    lock_guard: RwLockWriteGuard<'a, BTreeMap<ItemId, AuctionBiddingState>>,
+}
+
+impl<'a> TransactionImpl<'a> for InMemoryTransaction<'a> {
+    fn commit(self: Box<Self>) -> Result<()> {
+        Ok(())
+    }
+
+    // TODO: simulating rollbacks in a general way is not trivial
+    // and it would require all the `InMemory*` stores implementations
+    // to register previous value when creating the transaction or
+    // something like this.
+    fn rollback(self: Box<Self>) -> Result<()> {
+        bail!("Not supported")
+    }
+
+    fn load_tr(&mut self, item_id: ItemIdRef) -> Result<Option<AuctionBiddingState>> {
```

Comment:
OK. Here this approach is falling short, IMO.

I am not an expert in how usually this is modeled, but I was hoping that I will not have put every method of every possible store, on `impl TransactionImpl for SomeConcreteTransactionType`. 

I don't even think it's practical to come up with one unified interface for all possible databases and such.

I was hoping that I can implement the exact needed method on each `impl StoreType for ConcreteStore` somehow, and that mean I need to get the concrete type of the `Transaction` in there.

I am aware that moving around that `P` is a PITA, and was hoping that maybe some `Any` casting will be a solution here.

I need to think a bit more about it. I will try out with ideas from here.

Reply:
Agree that this is a point with some tradeoffs! 

I have two thoughts here

>but I was hoping that I will not have put every method of every possible store, on impl TransactionImpl for SomeConcreteTransactionType.

I think it's useful to reflect on "why" here. This does mean that you'd have to implement the entirety of database access layer in one module. Basically, every SQL query would have to go in a single file. That obviously isn't scalable for large applications. 

But, on the other hand, having all SQL queries specified in a single file is kinda cool? You get a physical reification of "application/database" boundary. 

So, the first though is that, Ifor the projects of auction sniper scale, I'd probably just go with this simple approach. 

The second thought is that we can split the interface in parts, and have one place that composes them together (instead of one place that implements the thing). Something along these lines:

```rust
pub struct Transaction(Box<dyn TransactionImpl>);

pub trait BidTransaction {
    fn create_bid(&mut self);
}

pub trait AuctionTransaction {
    fn create_auction(&mut self);
}

impl BidTransaction for Transaction {
    fn create_bid(&mut self) {
        self.0.create_bid()
    }
}
impl AuctionTransaction for Transaction {
    fn create_auction(&mut self) {
        self.0.create_auction()
    }
}

trait TransactionImpl: BidTransaction + AuctionTransaction {
    fn commit(self: Box<Self>);
    fn rollback(self: Box<Self>);
}
```

The problem here is that you need to keep everything in a single crate, and, if you are ow with doing that, than maybe the flat solution would work as well. 

I think, to allow writing `impl BidTransaction for Transaction` in the _downstream_ crate we'd need to have some amount of dynamic casting there. This won't get A+ in the "expressive problems" class, but would be ok for "pragmatic programmer".

I think this doesn't need Any actually, you can have

```rust
trait TransactionImpl {
    fn as_memory(&mut self) -> Option<&mut MemoryTx> { None }
    fn as_postgress(&mut self) -> Option<&mut PostgressTx> { None }
}
```

if you are willing to put both postgress and memory defintions into the same crate that defines `Transaction`. You need `Any` only if you go fullly decopuled way, where you have the following crates

```
tx
mem_tx depends on tx
pg_tx depends on tx
bid depends on tx
pg_bid depends on tx & bid & pg_tx
mem_bid depends on tx & bid & mem_tx

main depends on all of that
```

Reply:
Thanks. I'm going to need more time to think about this. BTW. I'm very much OK with pragmatic tradeoffs, though as this is a research exercise, I'm looking for most generic solution applicable to general scale. Whomever might use this as an example in the future (at very least myself), will have easier time adding simplifications to more complex&general solution, than the other way around.

Reply:
After grieving for a couple of evenings, I've dropped the static polymorphism altogether, and just converted to dynamic one as the only option. I couldn't sanely reconsile them and I agree that this `<P>` is a terrible UX.

RN I use type aliases, but I think I will convert to newtypes when I gain some confidence.

```
trait TransactionImpl {
    fn as_memory(&mut self) -> Option<&mut MemoryTx> { None }
    fn as_postgress(&mut self) -> Option<&mut PostgressTx> { None }
}
```

would definitely work, but I would like to challenge this to get the most general case, so I want that down-casting.

Using `Any` works without problem for `dyn Connection`, since it's `: 'static`, but for `dyn Transaction<'a>` I'm hitting the limitation `Any`. My understanding is that `trait Any : 'static` is purely because reasoning about casting lifetimes is complex in a general case, so it was never implemented.

I wrapped everything in some maybe-hopefully-safe-abstraction https://github.com/dpc/sniper/blob/beeb5cf150be48bfe2e5b55341bc2d5049d53205/src/persistence.rs#L57 , but it's above my Rustonomicon-fu level to know if it's actually safe.

This is where `unsafe` happens: https://github.com/dpc/sniper/blob/beeb5cf150be48bfe2e5b55341bc2d5049d53205/src/persistence/in_memory.rs#L66

And will of these, the Store implementations can somewhat neatly get the concrete type: https://github.com/dpc/sniper/blob/beeb5cf150be48bfe2e5b55341bc2d5049d53205/src/service/bidding_engine/postgres.rs#L17

I guess I'll go around the community and ask for some opinions on the safety of all this. Even if it's currently unsound, I hope it can be made into a sound thing.